Add support for internal nodes

KoBeWi opened the pull request:
Internal nodes are nodes that aren't exposed to the user. This is useful for control nodes that use other control nodes and we don't want anyone to meddle with them (e.g. ColorPicker or WindowDialog).

Compared to normal nodes:

- you can't normally access them with `get_child()` or `get_children()`
- `get_child_count()` will not count internal nodes, `get_position_in_node()` will return
position - internals size
- `move_child()`'s position will skip these nodes too
- you can however access them by passing `false` to `skip_internal` argument of aforementioned methods
- `skip_internal` is `true` by default in bindings (i.e. scripts), but it's false in code (so anything relying on children indexes won't break)
- internal nodes are always at the beginning of children list, even when added after normal nodes

This is supposed to fix #30381, close #21458 and resolve #7142
(also fixes #52146, but the issue might as well get a dedicated PR)

It's probably also useful for ppl making plugins. You can make a node internal by passing `true` to the new third argument of `add_child()`.

~~This is still WIP. While implementation is ok, internal nodes should be added to other controls (right now it's only ColorPicker) and it needs doc update.~~

TheDuriel commented:
Perhaps an interesting tidbit: Last i tried the internal nodes of GraphEdit show up in the scene tree, and can be modified. (Like hiding them to prevent users from changing the zoom settings.) But this also totally breaks GraphEdit.

girng commented:
I played around with this after looking at the issues linked in the OP.  

As mentioned in the OP, they are not supposed to be meddled with.  They are **vital** for their parent Node type (RTL, WindowDialog, etc) to _function_.  Should Godot crash? No, but I think those are specific cases that need to be dealt with. 

 A developer can simply create a node and use it as a container to put their nodes in. Then `queue_free` the children in there. That is.. If they really want to loop through the children and remove them.  The developer can also call `queue_free` on the parent node, and it will not cause the crash.

The developer can also use [leezh's example](https://github.com/godotengine/godot/issues/7142#issuecomment-261756979). 

Also, some GUI Nodes it's quite obvious they have internal nodes. (RichTextLabel's scrollbar in the editor)

Maybe an option to toggle the internal nodes in the editor hierarchy list could help. The real solution though  IMO, is for Godot not to silently crash when an internal node is free'd. I don't know how to even come up with a solution for that, though.

mrcdk commented:
I'm, personally, against this. Internal nodes are still nodes and their access should be the same as any other node. You can already distinguish internal nodes from other nothes because those internal nodes' `owner` property will be `null`

 I think this is more of a documentation issue. 

If anything, the only modification I'd do is to change `Node.get_children()` to `Node.get_children(with_internal_children = false)` and only return nodes with an `owner` property if `with_internal_children` is `true`.

TheDuriel commented:
Nodes added at runtime do not have an owner.

KoBeWi commented:
> If anything, the only modification I'd do is to change Node.get_children() to Node.get_children(with_internal_children = false)

Actually, it could be `true` by default and `false` in bindings (if that's possible).

> The real solution though IMO, is for Godot not to silently crash when an internal node is free'd.

That means null checks everywhere the crash might happen. It would negatively impact performance (not much, but still) and failed nodes are useless anyways.

The idea of this PR is to make some nodes less obvious to access and more importantly - impossible to access accidentally, which only leads to confusion and unexpected behavior.

KoBeWi commented:
I just pushed a commit that added `skip_internal` parameter to (hopefully) all methods that use node indexes. In code it's `false` by default and in bindings it's `true`.

Here's a tiny project for comparison of how `get_children` etc. behaves with ColorPicker:
[ColorPickerTest.zip](https://github.com/godotengine/godot/files/3365786/ColorPickerTest.zip)
Try it with 3.1 and this PR.

What's left to do is:
- add internals to other Controls that should use them
- update docs

aaronfranke commented:
I'm against completely hiding these, as it would mean hiding functionality from the user. If implemented, there should definitely be some kind of "get_child_unsafe" method or something.

It might be a better idea to allow users to see these nodes in the editor, which would mean that adding a node would generate its children, instead of generating them when the project is run.

> Turns out it's not straight-forward to use this option, because many nodes rely on accessing children, like Containers and stuff. Sooo, I thought of 2 ways to solve this:

For nodes with fixed children, would it be better to store a pointer to said child node, or connecting the nodes together with signals, so that it doesn't rely on getting children?

KoBeWi commented:
> I'm against completely hiding these, as it would mean hiding functionality from the user. If implemented, there should definitely be some kind of "get_child_unsafe" method or something.

Yeah, that's why I made it "hidden by default" with my last commit. The `skip_internal` argument allows you to access these nodes when set to `false`. I'll edit PR description to mention this.

> It might be a better idea to allow users to see these nodes in the editor, which would mean that adding a node would generate its children, instead of generating them when the project is run.

ColorPicker has 7 internal children. I don't think it's a good idea to show them in editor. There's also _zero_ need to edit them normally.

> For nodes with fixed children, would it be better to store a pointer to said child node, or connecting the nodes together with signals, so that it doesn't rely on getting children?

But it wasn't about nodes with fixed children. Container doesn't know what you put in it beforehand. Well of course, we could tell it explicitly what nodes it has underneath or it could remember them after doing `add_child`, but that sounds more complicated to work with.

aaronfranke commented:
> But it wasn't about nodes with fixed children. 

What I mean is how ColorPicker has 7 internal children. AFAIK these don't change.

KoBeWi commented:
The thing is, ColorPicker already stores pointers to them and its internal logic is perfectly fine (that's how you should use internal nodes; you don't need to access them even from the node that use them). But it's also a BoxContainer and we would have to change the Container implementation to support "fixed nodes". Not sure if it's worth the effort and it's not really relevant to the PR, as you need to do something about these nodes either way (and I remind that right now you can access them totally accidentally and crash your app, which is not desired probably).

KoBeWi commented:
I updated the docs and added internal node usage to `WindowDialog`, `AcceptDialog`, `FileDialog`, `GradientEdit`, `GraphEdit`, `ItemList`, `LineEdit`, `MenuButton`, `OptionButton`, `PopupMenu`, `RichTextLabel`, `ScrollContainer`, `Spinbox`, `TabContainer`, `TextEdit` and `Tree` (they were all relying on hidden nodes and potentially had the same unexpected problems). Marking this as ready for review. I'll squash the commits when it gets approved.

pycbouh commented:
While I completely support this PR as a solution for my own issue #38714, I'm still concerned about silent crashes. Can the crashes at least be "loud", leave some trace behind? If something is trying to access a freed instance, wouldn't there be an error? Is it simply suppressed?

Alternatively, what if `remove_child` or `queue_free` or both don't work on nodes marked as `internal` by default and require an optional argument to be passed that would force them to work? That way you wouldn't be able to accidentally cause a crash and would have to think about the call you are making. But, I guess, it would complicate things a lot internally.

KoBeWi commented:
I rebased the PR, hopefully everything still works 🙃

LikeLakers2 commented:
If you don't mind me suggesting, could there perhaps be a way for a GDScript user to check if a node is marked as "internal"? Perhaps a `Node.internal: bool` property, or `Node.is_internal() -> bool` function?

TheDuriel commented:
A meta data field would probably suffice, the editor already uses those for things like these.

KoBeWi commented:
> I also think such a "internal" setting might be useful to expose to the Editor Inspector. But that's just me.

The thing is that internal nodes don't show up in the scene tree (unless remote), so it wouldn't be useful.
`is_internal()` method would be easy to add.

reduz commented:
I am not happy with this solution because the problem is not clear. For the most part you can check whether a node is owned or not while editing or opening a loaded scene and this should be enough. If the problem sums up to "i am going around nodes that I may have added myself and I am not sure if I added them myself" then the problem is definitely not clear.

YeldhamDev commented:
@reduz One big problem that this PR solves is the implementation of things like godotengine/godot-proposals#2514. Nodes (be it from the engine itself or made by others) that rely on children being added to then, but also have nodes themselves would become much more easier to implement this way.

reduz commented on scene/main/node.cpp:
```diff
@@ -2741,13 +2756,13 @@ void Node::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_owner"), &Node::get_owner);
 	ClassDB::bind_method(D_METHOD("remove_and_skip"), &Node::remove_and_skip);
-	ClassDB::bind_method(D_METHOD("get_index"), &Node::get_index);
+	ClassDB::bind_method(D_METHOD("get_index", "skip_internal"), &Node::get_index, DEFVAL(true));
```
A comment here should be added that this is reversed from the C++ binding on purpose. This is super clever, but I did not notice it myself at first.

reduz commented on scene/main/node.cpp:
```diff
@@ -1864,8 +1875,12 @@ void Node::_propagate_replace_owner(Node *p_owner, Node *p_by_owner) {
 
-int Node::get_index() const {
-	return data.pos;
+int Node::get_index(bool p_skip_internal) const {
```
This is troublesome because the node may be an internal node and you are asking t to skip itself. It should be checked somewhere and throw an error.

reduz commented on scene/main/node.cpp:
```diff
@@ -1237,7 +1240,7 @@ void Node::add_child(Node *p_child, bool p_legible_unique_name, bool p_internal)
 
 	_add_child_nocheck(p_child, p_child->data.name);
```
The problem here is that you are assuming all internal nodes will come after but, in some situations, this is not desired (such as scrollbars above children controls in ScrollContainer). This system may work if you specify where you want internal children behind or after the user nodes, so you might need two separate arrays.

KoBeWi commented:
I pushed some changes. There is now distinction between "front" internal nodes and "back" internal nodes. I removed `internal_nodes` list as it wasn't really used. The internal nodes still need to be on the main list for everything else to work, so it was only some weird "mirror" and I only used the size of that list. Instead, now Node has two variables: `internal_children_front` and `internal_children_back`. They are modified whenever internal node is added or removed. Anything with index below `internal_children_front` is considered "front internal node", and analogously with the other variable. Instead of boolean, `add_child()` now uses an integer. Less than 0 means "front internal node", greater than 0 means "back internal node".

Now this all sounds very confusing, but luckily a regular user will almost never need to use the internal nodes (outside core, it's maybe useful for plugins), so GDScript API isn't any more complex than before. But internally, Node needs to make sure that a child doesn't move beyond its expected index (i.e. if node is "front internal child", it can't be moved past `internal_children_front` index and opposite - regular node can't be moved before this index, otherwise things won't work as expected. It won't crash though, so at least there's that)

Also I changed `skip_internal` parameters to `include_internal`.

The PR still needs rebase, testing, doc update etc. so I'm putting it as draft for now.

KoBeWi committed:
Add support for internal nodes

reduz commented on scene/main/node.cpp:
```diff
@@ -1070,7 +1107,7 @@ void Node::_add_child_nocheck(Node *p_child, const StringName &p_name) {
 }
 
-void Node::add_child(Node *p_child, bool p_legible_unique_name) {
+void Node::add_child(Node *p_child, bool p_legible_unique_name, int p_internal) {
```
Additionally, I am uncertain how many times legible unique name is used (I think only the editor uses it), so it may be better to swap the order of the two arguments to make this one come second.

kleonc commented:
> I need to somehow keep track of how many internal nodes of each type are there, so the indices can't change. They could be renamed though, but the name I used makes sense I think.

@KoBeWi I've actually misunderstood what `internal_children_back` is (now I've looked at the code :smile:). I thought it's an index and `internal_children_back == size() - internal_children_back_count` since you've said:
> Anything with index below `internal_children_front` is considered "front internal node", and analogously with the other variable.

But it's actually the count of back internal nodes.
Anyway, since these variables are used only internally (!) in the `Node` their names aren't that important so they're probably fine as they are. On the other hand, for the same reason, longer names shouldn't be a problem. Thus I'd change them to something like `internal_children_count_front`, `internal_children_count_back`. I find them clearer. But as said, it's not a big deal, it's up to you.

reduz commented:
Looks good to me, great job!

YeldhamDev commented:
I think the PR is great as it is, but I also agree with reduz that the `legible_unique_name` and `internal` arguments could be swapped.

KoBeWi commented:
> I also agree with reduz that the legible_unique_name and internal arguments could be swapped.

What? Where he said that?
Also `internal` has very little use outside core. It might be useful only for some plugins. It's better left at the very end of the argument list.

YeldhamDev commented:
@KoBeWi
>Additionally, I am uncertain how many times legible unique name is used (I think only the editor uses it), so it may be better to swap the order of the two arguments to make this one come second.

KoBeWi commented:
Yep, I just saw it. `legible_unique_name` is useful when you create a new node (e.g. `Node2D.new()`) and don't want it to have random name, but also don't want to name it yourself. It can even be used during runtime, to recognize nodes on remote tree. It's not used often, but has more applications for regular user than `internal` parameter. I'd leave it as is.

reduz commented:
Yeah I agree I think its better as it is now, your point of not making users pass this internal argument unnecesary makes more sense than legible unique name.
